contracts/bonding/BondingVotes.sol

```diff
@@ -0,0 +1,534 @@
+            bondedAmount: _bondedAmount,
+            delegateAddress: _delegateAddress,
+            delegatedAmount: _delegatedAmount,
+            lastClaimRound: _lastClaimRound,
+            lastRewardRound: _lastRewardRound
+        });
+        checkpoints.data[_startRound] = bond;
+
+        // now store the startRound itself in the startRounds array to allow us
+        // to find it and lookup in the above mapping
+        checkpoints.startRounds.pushSorted(_startRound);
+
+        onCheckpointChanged(_account, previous, bond);
+    }
+
+    function onCheckpointChanged(
+        address _account,
+        BondingCheckpoint memory previous,
+        BondingCheckpoint memory current
+    ) internal {
+        address previousDelegate = previous.delegateAddress;
+        address newDelegate = current.delegateAddress;
+        if (previousDelegate != newDelegate) {
+            emit DelegateChanged(_account, previousDelegate, newDelegate);
+        }
+
+        bool isTranscoder = newDelegate == _account;
+        bool wasTranscoder = previousDelegate == _account;
+        if (isTranscoder) {
+            emit DelegateVotesChanged(_account, previous.delegatedAmount, current.delegatedAmount);
+        } else if (wasTranscoder) {
+            // if the account stopped being a transcoder, we want to emit an event zeroing its "delegate votes"
+            emit DelegateVotesChanged(_account, previous.delegatedAmount, 0);
+        }
+
+        // Always send delegator events since transcoders are delegators themselves. The way our rewards work, the
+        // delegator voting power calculated from events will only reflect their claimed stake without pending rewards.
+        if (previous.bondedAmount != current.bondedAmount) {
+            emit DelegatorVotesChanged(_account, previous.bondedAmount, current.bondedAmount);
+        }
+    }
+
+    /**
+     * @notice Returns whether an account already has any checkpoint.
+     * @dev This is meant to be called by a checkpoint initialization script once we deploy the checkpointing logic for
+     * the first time, so we can efficiently initialize the checkpoint state for all accounts in the system.
+     */
+    function hasCheckpoint(address _account) public view returns (bool) {
+        return bondingCheckpoints[_account].startRounds.length > 0;
+    }
+
+    /**
+     * @notice Called by the BondingManager when the total active stake changes.
+     * @dev This is called only from the {BondingManager-setCurrentRoundTotalActiveStake} function to set the total
+     * active stake in the current round.
+     * @param _totalStake From {BondingManager-currentRoundTotalActiveStake}
+     * @param _round The round for which the total active stake is valid. This is normally the current round.
+     */
+    function checkpointTotalActiveStake(uint256 _totalStake, uint256 _round) public virtual onlyBondingManager {
+        if (_round > clock()) {
```

Comment:
Should this be `_round != clock() + 1` to mirror the conditional in `checkpointBondingState()`?

Reply:
I've changed this to `_round != clock()` and renamed the error to `InvalidTotalStakeCheckpointRound`.

The checkpointing rounds are actually different between individual accounts and total active stake tho. While individual accounts are always checkpointed on the next round, from the way that active stake update works on BondingManager, the total active stake is checkpointed for the current round instead, when the round is initialized and `nextRoundTotalActiveStake` becomes the current.

We can see it as having an optimization for the total active stake: since we have that single point where we update 1 variable for the total stake, we can leave the checkpointing for that time only. That is cheaper than checkpointing the `nextRoundTotalActiveStake` for the next round every time it changes, which would be the analogous flow to the individual accounts checkpoints.